README.md
# traq

![travis build status](https://travis-ci.org/nicolai86/traq.svg?branch=master)

time tracking using text files

## Usage examples

``` bash
# start time tracking for #project
$ traq project

# start time tracking for #project with the comment 'working on the landing page'
$ traq project working on the landing page

# stop time tracking
$ traq stop

# echo the content of todays file to stdout. If the file does not exist, nothing is echoed.
$ traq

# echo the content of the file from the given date to stdout. If the file does not exist, nothing is echoed.
$ traq -d 2012-07-30

# echo the content of all files in february of the current year
$ traq -m 2

# echo the content of all files in september 2012
$ traq -m 9 -y 2012

# starts time tracking for client-a on #development
$ traq -p client-a development

# stops time tracking for client-a
$ traq -p client-a stop

# list tracked times for client-a from today
$ traq -p client-a
```

## Evaluation

To evaluate traq files pass the `-e` command line flag to the utility.
E.g.

    $ traq -p test -e
    2012-09-27
    #foo:0.1666
    #bar:0.1666
    %%

## Installation from source

``` bash
$ go get github.com/nicolai86/traq
$ echo "export TRAQ_PATH=$GOPATH/src/github.com/nicolai86/traq" >> ~/.bash_profile
$ echo "export TRAQ_DATA_DIR=$HOME/Library/traq" >> ~/.bash_profile
$ ln -s $TRAQ_PATH/man/traq.1 /usr/local/share/man/man1/traq.1
$ . ~/.bash_profile
$ which traq
```

### Build instructions

Make sure your `$GOPATH` is set up properly. Then link traq properly:

``` bash
$ mkdir -p $HOME/go/src
$ ln -s $HOME/.traq/src/ $HOME/go/src/traq
```

``` bash
$ go build -o bin/traq app.go
```

### Go Formatting

``` bash
$ gofmt -w traq.go
```
## Bash Completion

If you have `bash-completion` installed you can setup bash completion for traq as well. This example assumes you are using [HomeBrew][1] and have `bash-completion` installed.

``` bash
$ ln -s $TRAQ_PATH/traq_completion.sh $(brew --prefix)/etc/bash_completion.d/traq
```

Ubuntu users can do the following:

``` bash
$ sudo apt-get install bash-completion
$ echo ". $TRAQ_PATH/traq_completion.sh" >> ~/.bash_profile
```

## Migration to v0.5

traq 0.5 has a different, flatter directory structure. Instead of one directory per year week of the year,
we now only have one directory per year.

The following bash script helps you migrate your data:

```bash
for directory in $(find $HOME/Library/traq -maxdepth 1 -mindepth 1 -type d); do
  echo $directory
  for year in $(find $directory -maxdepth 1 -mindepth 1 -type d); do
    for week in $(find $year -maxdepth 1 -mindepth 1 -type d); do
      for file in $(find $week -maxdepth 1 -mindepth 1 -type f); do
        cleanfile="${file//timestamps-/}"
        mv "$file" "$year/${cleanfile##*/}"
      done
      rm -fr $week
    done
  done
done
```

## Hacking

All files are placed under

    $ $TRAQ_DATA_DIR/timestamps/<current year>/<year>-<month>-<day>
    # eg $TRAQ_DATA_DIR/timestamps/2012/2012-12-12

or, if `-p <project>` was given, under

    $ $TRAQ_DATA_DIR/<project>/<current year>/<year>-<month>-<day>
    # eg $TRAQ_DATA_DIR/client-a/2012/2012-12-12

Each file can contain multiple lines of the following format:


    <timestamp>;<tag>;<comment>


Here's some sample content:

    Thu Sep 27 07:05:05 +0400 2012;#foo;Worked on Foo
    Thu Sep 27 07:15:05 +0400 2012;#bar;
    Thu Sep 27 07:25:05 +0400 2012;stop;

[1]:http://mxcl.github.com/homebrew/


CHANGELOG
# v0.7

- add support to summarize given time ranges, e.g. summarize all times within a month, using the -s flag

# v0.6

- running timer evaluation using -r flag

# v0.5

- remove -w flag which printed times for a given week
- change directory structure of stored timestamps
- rewrite traq in go
- introduce -e flag to perform evaluations inside of traq

# v0.4

- automatically cancel trailing tags to ensure that traqeval does not calculate invalid
  durations

# v0.3

- traq now supports month and year batch printing of tracked times
  use -m to provide a month and -y to provide a year.
  Both -m and -y have support for bash completion, based on the current year
  and month.

# v0.2.1

- traqtrans works with pipes or file as first argument
- add tests for traqtrans and traqeval

# v0.2

- add support for comments

# v0.1

- add bash completion support
- add customizable data directory (@passy)

# v0.0.2

- traqeval is based on bash now
- add tests for traq_trans

# v0.0.1

- add support to track times using a simple text based format
- add support to track times for different projects using the `-p <project>` switch
- add support to list the current days tracking using an empty argument list
- add support to list the times for a week using the `-w <weeknumber>` switch
- add `traqtrans` programm to convert ISO 8601 timestamps to unix timestamps
- add `traqeval` programm to sum unix timestamps

traq.go
// Copyright 2013-2014 Raphael Randschau

/*
Package traq implements a CLI for time tracking.
*/
package main

import (
	"flag"
	"fmt"
	"os"
	"time"
)

// DatesInMonth calculates the days of a given month and year.
func DatesInMonth(year int, month int) []time.Time {
	var dates []time.Time = make([]time.Time, 0)
	var date time.Time = time.Date(year, time.Month(month), 1, 0, 0, 0, 0, time.UTC)

	for {
		dates = append(dates, date)
		date = date.Add(time.Hour * 24)
		if int(date.Month()) != month {
			break
		}
	}

	return dates
}

// SumFile evaluates the content of a traq tracking file.
// The returned map contains every tag contained in the file as well as the
// tracked duration in seconds.
func SumFile(lines []TimeEntry) (map[string]int64, error) {
	var totalled map[string]int64 = make(map[string]int64)

	var currentTag string = ""
	var currentTime time.Time

	for _, entry := range lines {
		_, ok := totalled[entry.Tag]
		if !ok {
			totalled[entry.Tag] = 0
		}

		if entry.Tag == "stop" {
			var diff = entry.Date.Unix() - currentTime.Unix()
			totalled[currentTag] = totalled[currentTag] + diff
			currentTag = ""
		} else if currentTag != entry.Tag {
			var diff = entry.Date.Unix() - currentTime.Unix()
			totalled[currentTag] = totalled[currentTag] + diff
			currentTag = entry.Tag
		}

		currentTime = entry.Date
	}

	// we did not end with a stop tag
	// this problem can be avoided by using the RunningEvaluator when parsing a time file
	if currentTag != "stop" {
		var diff = time.Date(currentTime.Year(), currentTime.Month(), currentTime.Day(), 23, 59, 59, 0, currentTime.Location()).Unix() - currentTime.Unix()
		totalled[currentTag] = totalled[currentTag] + diff
		currentTag = ""
	}

	delete(totalled, "")
	delete(totalled, "stop")

	return totalled, nil
}

// PrintDate prints the content of a single traqfile, identified by the project identifer
// and its date
func PrintDate(storage TimeEntryReader, dates ...time.Time) {
	for _, date := range dates {
		content, err := storage.Content(date)

		if err == nil {
			for _, entry := range content {
				fmt.Printf("%s;%s;%s\n", entry.Date.Format("Mon Jan 02 15:04:05 -0700 2006"), entry.Tag, entry.Comment)
			}
			fmt.Println("%%")
		}
	}
}

func SummarizeDate(storage TimeEntryReader, dates ...time.Time) map[string]int64 {
	var tags map[string]int64 = make(map[string]int64)
	for _, date := range dates {
		content, err := storage.Content(date)

		if err == nil {
			var totalled, _ = SumFile(content)

			for key, value := range totalled {
				current, ok := tags[key]
				if !ok {
					tags[key] = value
				} else {
					tags[key] = value + current
				}
			}
		}
	}

	return tags
}

func PrintSummary(tags map[string]int64) {
	for key, value := range tags {
		fmt.Printf("%s:%2.4f\n", key, float64(value)/60.0/60.0)
	}
}

// EvaluateDate prints the evaluation of a single traqfile, identified by the project identifier
// and its date
func EvaluateDate(storage TimeEntryReader, dates ...time.Time) {
	for _, date := range dates {
		var content, error = storage.Content(date)

		if error == nil {
			fmt.Printf("%d-%02d-%02d\n", date.Year(), date.Month(), date.Day())
			var totalled, _ = SumFile(content)
			// TODO handle errors
			for key, value := range totalled {
				fmt.Printf("%s:%2.4f\n", key, float64(value)/60.0/60.0)
			}

			fmt.Println("%%")
		}
	}
}

func main() {
	var (
		month    = flag.Int("m", 0, "print tracked times for a given month")
		year     = flag.Int("y", 0, "print tracked times for a given year")
		project  = flag.String("p", "timestamps", "print data for a given project")
		date     = flag.String("d", "", "print tracked times for a given date")
		evaluate = flag.Bool("e", false, "evaluate tracked times")
		running  = flag.Bool("r", false, "add fake stop entry to evaluate if stop is missing")
		summary  = flag.Bool("s", false, "summaries the given timeframe")
	)

	flag.Parse()

	var now = time.Now()
	var t, error = time.Parse("2006-01-02", *date)
	if error == nil {
		*year = t.Year()
		*month = int(t.Month())
	} else {
		if *year == 0 {
			*year = now.Year()
		}
		if *month == 0 {
			*month = int(now.Month())
		}
	}

	var loader LogLoader = ContentLoader

	if *running {
		loader = RunningLoader
	}

	storageProvider := FileSystemStorage{os.Getenv("TRAQ_DATA_DIR"), *project, loader}

	if *evaluate {
		if *date == "" {
			EvaluateDate(&storageProvider, DatesInMonth(*year, *month)...)
		} else {
			EvaluateDate(&storageProvider, t)
		}
		return
	}

	var command string = flag.Arg(0)

	printDate := !*summary
	handler := func(storage FileSystemStorage, dates ...time.Time) {
		if printDate {
			PrintDate(&storageProvider, dates...)
		} else {
			PrintSummary(SummarizeDate(&storageProvider, dates...))
		}
	}

	if command == "" {
		if *date == "" {
			handler(storageProvider, DatesInMonth(*year, *month)...)
		} else {
			handler(storageProvider, t)
		}
	} else {
		storageProvider.Store(TimeEntry{now, command, ""})
	}
}
